src/shared/components/cnSegments/CNSegmentsDownloader.tsx

```diff
@@ -0,0 +1,74 @@
+import * as React from "react";
+import MobxPromise from "mobxpromise/dist/src/MobxPromise";
+import {observable} from "mobx";
+import {observer} from "mobx-react";
+import {Button} from "react-bootstrap";
+import classnames from "classnames";
+import DefaultTooltip from "shared/components/defaultTooltip/DefaultTooltip";
+import {CopyNumberSeg} from "shared/api/generated/CBioPortalAPI";
+import {generateSegmentFileContent} from "shared/lib/IGVUtils";
+import onMobxPromise from "shared/lib/onMobxPromise";
+import fileDownload from "react-file-download";
+import autobind from "autobind-decorator";
+
+export type CNSegmentsDownloaderProps = {
+    promise: MobxPromise<CopyNumberSeg[]>;
+    filename?: string;
+    buttonClassName?: string;
+    tooltipPlacement?: string;
+};
+
+@observer
+export default class CNSegmentsDownloader extends React.Component<CNSegmentsDownloaderProps, {}> {
+    public static defaultProps = {
+        filename: "segments.seg",
+        buttonClassName: "btn btn-sm pull-right",
+        tooltipPlacement: "left"
+    };
+
+    @observable downloading = false;
+    private downloadFileContent: string|undefined;
+
+    public render() {
+        return (
+            <DefaultTooltip
+                overlay={<span>Download a copy number segment file for the selected samples</span>}
+                placement={this.props.tooltipPlacement}
+            >
+                <Button
+                    className={this.props.buttonClassName}
+                    onClick={this.handleDownload}
+                    disabled={this.downloading}
+                >
+                    <i className={classnames({'fa': true,
+                        'fa-cloud-download': !this.downloading,
+                        'fa-spinner fa-pulse': this.downloading})}
+                    />
+                </Button>
+            </DefaultTooltip>
+        );
+    }
+
+    @autobind
+    private handleDownload() {
+        if (this.downloadFileContent) {
+            fileDownload(this.downloadFileContent, this.props.filename);
+        }
+        else if (this.props.promise.isComplete) {
```

Comment:
i think you could just use onMobxPromise once and it would handle case where it promise is resolved or still pending

Reply:
If the promise is already complete I didn't want to set `this.downloading` to true upon download button click, otherwise the `else` block below is already using `onMobxPromise`. I'll try to simplify & remove the additional `else if` block.

Reply:
Just tried to remove this `else if` block, it has virtually no effect on the loading state of the button. I think it is safe to remove it.